### PR TITLE
Aggregrate properties for composite allOf types

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Don't anonymously embed composite types, promote their properties
into the containing type.  This avoids unnecessary levels of nesting
and represents what's sent over the wire.  It also simplifies the
handling of custom marshallers and unmarshallers.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/autorest.go/issues/732

File: src/transform/transform.ts [hunk 1]
```diff
@@ -72,15 +72,22 @@ async function process(session: Session<CodeModel>) {
...
-          const asObj = <ObjectSchema>child;
-          let discValue = getEnumForDiscriminatorValue(asObj.discriminatorValue!, discriminatorEnums);
+        if (obj.discriminatorValue) {
+          let discValue = getEnumForDiscriminatorValue(obj.discriminatorValue!, discriminatorEnums);
           if (!discValue) {
             // this happens when there are intermediate types in the hierarchy that aren't discriminators.
             // e.g. BaseType -> GenericProps -> DerivedType
             // here, GenericProps is just a collection of standard properties common to various objects.
             // M4 gives it a discriminatorValue of "GenericProps" that won't exist in the enum defined
             // in BaseType.  unclear if this is desirable, but for now we just embed that value directly.
```
Comment (jhendrixMSFT) [inline src/transform/transform.ts:82]: Turns out this is a bug in M4.  We will be able to remove this in a future version.

Comment (jhendrixMSFT) [reply src/transform/transform.ts]: M4 has been fixed and incorporated.

**Comment by jhendrixMSFT** [comment]
Also fixes https://github.com/Azure/autorest.go/issues/709

**Review by seankane-msft — APPROVED** [review]
This removes embedded structs, will this be a breaking change for management plane?

**Comment by jhendrixMSFT** [comment]
It will be.  However, we think the change is worth it as it simplifies the usage and is the same that we generate in track 1 which will ease migration.

**Comment by seankane-msft** [comment]
> It will be. However, we think the change is worth it as it simplifies the usage and is the same that we generate in track 1 which will ease migration.

I would agree with you, in that case LGTM

**Pull request merged**